release/models/system/openconfig-system-grpc-tunnel.yang

```diff
@@ -0,0 +1,348 @@
+module openconfig-system-grpc-tunnel {
+  yang-version "1";
+
+  namespace "http://openconfig.net/yang/system/grpc-tunnel";
+  prefix "oc-sys-grpc-tun";
+
+  import openconfig-extensions { prefix oc-ext; }
+  import openconfig-system { prefix oc-sys; }
+  import openconfig-network-instance { prefix oc-netinst; }
+  import openconfig-inet-types { prefix oc-inet; }
+
+  organization
+    "OpenConfig working group";
+  contact
+    "www.openconfig.net";
+
+  description
+    "This module adds configuration and operational state relating to gRPC
+    tunnel (https://github.com/openconfig/grpctunnel/) to invoke one or
+    more dialout TCP over gRPC tunnels to bypass situations in which
+    direct dialin gRPC communications are not feasible.";
+
+
+  oc-ext:openconfig-version "0.1.0";
+
+  revision 2022-07-08 {
+    description
+      "Initial revision.";
+    reference "0.1.0";
+  }
+
+  oc-ext:catalog-organization "openconfig";
+  oc-ext:origin "openconfig";
+
+  typedef target-type {
+    type enumeration {
+      enum UNKNOWN {
+        value 0;
+        description
+          "An Unknown/Invalid target type";
+      }
+      enum SSH {
+        value 22;
+        description
+          "SSH protocol target type";
+      }
+      enum NETCONF_SSH {
+        value 830;
+        description
+          "NETCONF over SSH target type";
+      }
+      enum OPENFLOW {
+        value 6653;
+        description
+          "OpenFlow protocol target type";
+      }
+      enum GNMI_GNOI {
+        value 9339;
+        description
+          "gNMI/gNOI target type";
+      }
+      enum GRIBI {
+        value 9340;
+        description
+          "gRIBI target type";
+      }
+      enum P4RT {
+        value 9559;
+        description
+          "P4 Runtime target type";
+      }
+    }
+    description
+      "Sets the TCP based protocols to be tunneled over the gRPC tunnel";
+    reference
+      "OpenConfig gRPC Tunnel:
+      https://github.com/openconfig/grpctunnel/blob/master/proto/tunnel/tunnel.proto";
+  }
+
+  grouping tunnel-server-config {
+    description
+      "Configuration and state parameters corresponding to an individual
+      gRPC tunnel server";
+
+    leaf name {
+      type string;
+      description
+        "The name of the gRPC tunnel server instance";
+    }
+
+    leaf enable {
+      type boolean;
+      description
+        "When set to true, the tunnel server endpoint is enabled and an
+        outbound connection is initiated";
+    }
+
+    leaf host {
+      type oc-inet:host;
+      description
+        "The IPv4/IPv6 address or hostname of the tunnel server";
+    }
+
+    leaf port {
+      type oc-inet:port-number;
+      description
+        "The port number of the tunnel server";
+    }
+
+    leaf transport-security {
```

Comment:
We've discussed this a bit in the past -- should we ever allow TLS to be turned off for this application? The 'lab testing' thing is "convenient" but it's also relatively easy to just generate self-signed certificates.

Do we ever expect that someone would run this without TLS in production?

Reply:
This actually presents a unique case where there could be more justification that just lab/testing in that an operator could choose to disable TLS on the tunnel but enforce for the grpc-server connection (or vice-versa) if you only want a single layer of TLS

Reply:
I'm reluctant to add this option as it might inadvertently send a signal that encryption is reasonable to disable.  I realize there is an overhead associated but that seems a worthwhile security requirement to me.

Reply:
Currently, this toggle exists for each `grpc-server` as well - default to `true` with the same blurb

If this should be removed to not promote disabling transport security whatsoever, I would then suggest that the grpctunnel specification mandate encryption (as well as network-element based gRPC service endpoints for gNMI, gRIBI, gNOI, etc..  if it will be removed for grpc-servers as well)